src/tiny_v2.rs

```diff
@@ -0,0 +1,311 @@
+        let class_name = namespace_value!(base, parts, namespace_named_index)
+            .ok_or(MappingError::MissingClassName)?;
+
+        let official_name = namespace_value!(base, parts, namespace_official_index);
+        let intermediary_name = namespace_value!(base, parts, namespace_intermediary_index);
+
+        *current_class_name = class_name.clone();
+        self.classes.insert(class_name, ClassMapping::new(official_name, intermediary_name, HashMap::new(), HashMap::new()));
+        Ok(())
+    }
+}
+
+const BASE_CLASS_MEMBERS_NAMESPACE_OFFSET: usize = 3;
+
+impl ClassMapping {
+    fn parse_class_members(
+        &mut self,
+        namespace_named_index: usize,
+        namespace_intermediary_index: usize,
+        namespace_official_index: usize,
+        parts: &Vec<&str>,
+        subsection_type: &&str,
+        descriptor: Arc<str>
+    ) -> Result<(), MappingError> {
+        let named_name = namespace_value!(member, parts, namespace_named_index)
+            .ok_or(MappingError::MissingFieldOrMethodName)?;
+
+        let official_name = namespace_value!(member, parts, namespace_official_index);
+        let intermediary_name = namespace_value!(member, parts, namespace_intermediary_index);
+
+        if *subsection_type == METHOD_IDENT {
+            // Method section
+            self.methods.insert((named_name, descriptor), MethodMapping::new(official_name, intermediary_name));
+        } else {
+            // Field section
+            self.fields.insert((named_name, descriptor), FieldMapping::new(official_name, intermediary_name));
+        }
+        Ok(())
+    }
+}
+
+impl Mapping for TinyV2Mapping {
+    fn remap_class<C>(&self, class_name: C) -> Option<Arc<str>>
+    where
+        C: AsRef<str>
+    {
+        self.classes.get(class_name.as_ref())
+            .map(|class| class.official_name.clone().unwrap_or_else(|| Arc::from(class_name.as_ref())))
+    }
+
+    fn remap_method<C, M, D>(&self, class_name: C, method_name: M, descriptor: D) -> Option<Arc<str>>
+    where
+        C: AsRef<str>,
+        M: AsRef<str>,
+        D: AsRef<str>
+    {
+        let remapped_descriptor = self.remap_descriptor(descriptor.as_ref());
+        self.classes.get(class_name.as_ref())
+            .and_then(|class| class.methods.get(&(method_name.as_ref().into(), remapped_descriptor)))
+            .map(|method| method.official_name.clone().unwrap_or_else(|| Arc::from(method_name.as_ref())))
+    }
+
+    fn remap_field<C, F, D>(&self, class_name: C, field_name: F, descriptor: D) -> Option<Arc<str>>
+    where
+        C: AsRef<str>,
+        F: AsRef<str>,
+        D: AsRef<str>
+    {
+        let remapped_descriptor = self.remap_descriptor(descriptor.as_ref());
+        self.classes.get(class_name.as_ref())
+            .and_then(|class| class.fields.get(&(field_name.as_ref().into(), remapped_descriptor)))
+            .map(|method| method.official_name.clone().unwrap_or_else(|| Arc::from(field_name.as_ref())))
+    }
```

Comment:
Returning Arc<...> seems very annoying. Why are we doing this instead of cloning.